src/script/microbit-interfacing/serialProtocol.ts

```diff
@@ -50,40 +55,86 @@ export type MicrobitSensorState = {
 export const version = 1;
 
-// TODO: Revisit this regexes, they are probably not optimal
-const responseIdRegex = 'R\\[([\\w]*?)\\]';
-const handshakeRegex = /(?<=HS\[)[\d]+?(?=\])/;
-
-let commandId = 1;
-
 export const splitMessages = (message: string): SplittedMessages => {
   if (!message) {
     return {
       messages: [],
       remainingInput: '',
     };
   }
-  const messages = message.split('\n');
+  let messages = message.split('\n');
   let remainingInput = messages.pop() || '';
+
+  // Throw away any empty messages and messages that don't start with a valid type
+  messages = messages.filter(
+    (msg: string) =>
+      msg.length > 0 && Object.values(MessageTypes).includes(msg[0] as MessageTypes),
+  );
+
+  // Any remaining input will be the start of the next message, so if it doesn't start
+  // with a valid type throw it away as it'll be prepended to the next serial string
+  if (
+    remainingInput.length > 0 &&
+    !Object.values(MessageTypes).includes(remainingInput[0] as MessageTypes)
+  ) {
+    remainingInput = '';
+  }
+
   return {
     messages,
     remainingInput,
   };
 };
 
-export const processHandshake = (message: string): MessageResponse | undefined => {
-  const responseIdMatch = message.match(responseIdRegex);
-  if (!responseIdMatch) {
+export const processResponseMessage = (message: string): MessageResponse | undefined => {
+  // Regex for a message response with 3 groups:
+  // id    -> The message ID, 1-8 hex characters
+  // cmd   -> The command type, a string, only capital letters, matching CommandTypes
+  // value -> The response value, empty string or a word, number,
+  //          or version (e.g 1.2.3) depending on the command type
+  const responseMatch =
+    /^R\[(?<id>[0-9A-Fa-f]{1,8})\](?<cmd>[A-Z]+)\[(?<value>-?[\w.]*)\]$/.exec(message);
+  if (!responseMatch || !responseMatch.groups) {
     return undefined;
   }
-  let protocolVersion = Number(message.match(handshakeRegex)?.[0]);
-  if (!protocolVersion) {
+  const messageId = parseInt(responseMatch.groups['id'], 16);
+  if (isNaN(messageId)) {
     return undefined;
   }
+  const cmdType = responseMatch.groups['cmd'] as CommandTypes;
+  if (!Object.values(CommandTypes).includes(cmdType)) {
+    return undefined;
+  }
+  let value: string | number = responseMatch.groups['value'];
+  switch (cmdType) {
+    // Commands with numeric values
+    case CommandTypes.Handshake:
+    case CommandTypes.RadioFrequency:
+    case CommandTypes.HardwareVersion:
+      value = Number(value);
+      if (isNaN(value)) {
+        return undefined;
+      }
+      break;
+    // Commands without values
+    case CommandTypes.Zstart:
+    case CommandTypes.Stop:
+      if (value !== '') {
+        return undefined;
+      }
+      break;
+    // Semver-ish values (valid range 00.00.00 to 99.99.99)
+    case CommandTypes.SoftwareVersion:
+      if (!/^[0-9]{1,2}\.[0-9]{1,2}\.[0-9]{1,2}$/.test(value)) {
+        return undefined;
+      }
+      break;
```

Comment:
Do we need to support a semever-like version with more than this range? for example `v1.2.3-alpha.1`?

Reply:
I don't have context for this. What requirement is this addressing? Is it something we need to log or track in analytics?

Reply:
I think you've mentioned at some point that the BLE version gets the micro:bit programme version (via BLE UART?) and it uses that to determine if it needs to ask the user to update/reflash the micro:bit.
In the current implementation for the serial bridge we don't really need to do this, if we know the micro:bit has always been flashed, but in case it's useful to offer this feature to mirror the BLE features, there is a command to retrieve the "software version".

Reply:
For the moment we've disabled that code path for the Bluetooth code as it's buggy and we don't have the time to address properly. One to fix for when we have more than one bluetooth hex around (and after Rob's current work o the connect/disconnect process).

The full semver regexp is not that bad: https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string

The bluetooth C++ simply hardcodes "1" as its version and I suspect it may have been intended more as a protocol version but it's hard to be sure.

Reply:
Ah, interesting, once we look back into that we can figure out exactly what it is. If it's the protocol version that'd be equivalent to the number included with the handshake, if it's a version for the micro:bit programme then it would be this "software version" command.

About the regex, yeah, it should be fine to match it to the full semver format on the ML tool side, but for simplicity in the protocol description itself, I might restrict the "software version" to be always a a value between `0.0.0` and `99.99.99`, as we don't really have a need for more. Mostly I was wondering if maybe ml-machine was using semver (or any other version format) to do something a bit more unexpected, maybe something like appending the protocol version (e.g v1.0.0-protocol.1), but sounds like we don't need to worry.